docs/arcadia-users-group/20231128-testing-datasets/lesson.md

```diff
@@ -0,0 +1,289 @@
+This is more overhead than we typically strive for in a lesson, but we hope that it's a chance to practice these skills to achieve a new goal.
+In the future, we may provide a [GitPod](https://www.gitpod.io/) environment for learners to use while working through this lesson, however if possible we would prefer to empower users to start implementing tests on their own computers using their own setup.
+
+To start this lesson, we'll begin by creating a conda environment that has the tools we'll need.
+
+```
+mamba create -n augtest2 jupyter pytest pandas matplotlib seaborn pillow
+conda activate augtest2
+```
+
+We'll also create a folder to help us stay organized.
+```
+mkdir 20231128-aug-testing-lesson
+cd 20231128-aug-testing-lesson
+```
+
+Once our environment is activated, start a Jupyter notebook
+
+```
+jupyter notebook
+```
+
+### Example function
+
+As a running example, similar to the `mean()` function we used in the previous lesson, we'll use `distribution()`, a simple function that returns the distribution of characters in text.
+This function is intentionally implemented inefficiently, to emulate an expensive function of the type typically found in computational biology workflows, that would require testing.
+
+```python
+def distribution(text):
+    return { char: text.count(char)/len(text) for char in text.lower() }
+```
+
+Let's try it:
+
+```
+distribution('testing datasets')
+```
+
+```
+{'t': 0.25,
+ 'e': 0.125,
+ 's': 0.1875,
+ 'i': 0.0625,
+ 'n': 0.0625,
+ 'g': 0.0625,
+ ' ': 0.0625,
+ 'd': 0.0625,
+ 'a': 0.125}
+```
+
+## Test data for functions and scripts
+
+### Creating data within the test itself
+
+As seen in our previous lesson and described above, one strategy is to create test data within the test using built-in functions or data structures.
+When possible, this is a great strategy.
+These test data are small and self-contained and the correct answer to the test is easy to figure out.
+Functions like `[]` (list creation), or `range()` in Python and `c()`, `rep()`, `seq()`, `data.frame()`, and `list()` in R can be used to create test data that are relevant to a function.
+Some data-creation functions might be non-deterministic, but setting the seed ensures that the same output are produced each time.
+
+Often, we'll test a combination of "typical" inputs for which we verified the expected outputs, alongside edge cases - empty, very large, or nonsensical inputs.
+For `distribution()`, a typical in-test set of test cases could include:
+
+```python
+DISTRIBUTION_INPUTS_AND_EXPECTED_OUTPUTS = [
+    ('', {}),                                    # An empty string should produce an empty output
+    ('abcc', {'a': 0.25, 'b': 0.25, 'c': 0.5}),  # Typical short input
+    (100, None),                                 # Nonsensical input (wrong type)
+    ...
+]
```

Comment:
@giladmishne is this pseudo-set up for parameterized tests, or just examples of what reasonable test data looks like for this function?

Reply:
These are just examples for reasonable "inline" text cases.